### PR TITLE
boot: handle canceled update

### PR DESCRIPTION
**Pull request by bboozzoo**
When a trusted boot assets update gets canceled, we need to clean up the pending
changes we have accumulated in modeenv.

### CODE REVIEW

File: boot/assets.go [hunk 1]
```diff
@@ -539,11 +546,68 @@ func (o *TrustedAssetsUpdateObserver) BeforeWrite() error {
...
+				// asset was newly added
+				delete(*trustedAssets, changed.name)
+			}
+		} else {
+			(*trustedAssets)[changed.name] = hashList[:1]
+		}
+		if !isAssetHashTrackedInMap(otherTrustedAssets, changed.name, changed.hash) {
+			// asset revision is not used used elsewhere, we can remove it from the cache
+			if err := o.cache.Remove(changed.blName, changed.name, changed.hash); err != nil {
+				return fmt.Errorf("cannot remove unused boot asset %v:%v: %v", changed.name, changed.hash, err)
```
Comment (pedronis) [inline boot/assets.go:584]: this should probably be only logged, and then the function shouldn't return an error. This means below we always take care of both recovery and not recovery

File: boot/assets.go [hunk 2]
```diff
@@ -539,11 +546,68 @@ func (o *TrustedAssetsUpdateObserver) BeforeWrite() error {
...
+		return nil
+	}
+	for _, isRecovery := range []bool{false, true} {
+		if err := o.canceledUpdate(isRecovery); err != nil {
+			// XXX: should this be a log instead?
+			return err
+		}
+	}
+
+	if err := o.modeenv.WriteTo(""); err != nil {
```
Comment (pedronis) [inline boot/assets.go:606]: can't WriteTo be just Write?  (the same question applies to other places in the file)

File: boot/assets.go [hunk 3]
```diff
@@ -539,11 +546,68 @@ func (o *TrustedAssetsUpdateObserver) BeforeWrite() error {
...
+	if recovery {
+		trustedAssets = &o.modeenv.CurrentTrustedRecoveryBootAssets
+		otherTrustedAssets = o.modeenv.CurrentTrustedBootAssets
+		changedAssets = o.seedChangedAssets
+	}
+
+	for _, changed := range changedAssets {
+		if *trustedAssets == nil {
+			return nil
+		}
```
Comment (pedronis) [inline boot/assets.go:562]: should this if be before the for loop ?

File: boot/assets.go [hunk 4]
```diff
@@ -539,11 +546,68 @@ func (o *TrustedAssetsUpdateObserver) BeforeWrite() error {
...
+			if currentAssetHash != changed.hash {
+				// assets list has already been trimmed, nothing
+				// to do
+				continue
+			} else {
+				// asset was newly added
+				delete(*trustedAssets, changed.name)
+			}
+		} else {
+			(*trustedAssets)[changed.name] = hashList[:1]
```
Comment (pedronis) [inline boot/assets.go:579]: // update asset was last

**Review by pedronis — COMMENTED** [review]
thank you, did a pass

File: boot/assets.go [hunk 5]
```diff
@@ -575,16 +577,16 @@ func (o *TrustedAssetsUpdateObserver) canceledUpdate(recovery bool) error {
 				delete(*trustedAssets, changed.name)
 			}
 		} else {
+			// asset updates are appended to the list
```
Comment (pedronis) [inline boot/assets.go:580]: in this formulation maybe it's slightly clearer with s/are/were/ ?

**Review by pedronis — APPROVED** [review]
thank you

File: boot/assets.go [hunk 6]
```diff
@@ -455,7 +463,7 @@ func (o *TrustedAssetsUpdateObserver) observeUpdate(bl bootloader.Bootloader, re
 	if o.modeenv.deepEqual(modeenvBefore) {
 		return true, nil
 	}
-	if err := o.modeenv.WriteTo(""); err != nil {
```
Comment (anonymouse64) [inline boot/assets.go:458]: thanks for this

File: boot/assets_test.go [hunk 7]
```diff
@@ -1250,6 +1251,421 @@ func (s *assetsSuite) TestUpdateObserverUpdateRollbackGrub(c *C) {
...
+	err = obs.Canceled()
+	c.Assert(err, IsNil)
+	afterCancelM, err := boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	c.Check(afterCancelM.CurrentTrustedBootAssets, DeepEquals, m.CurrentTrustedBootAssets)
+	c.Check(afterCancelM.CurrentTrustedRecoveryBootAssets, DeepEquals, m.CurrentTrustedRecoveryBootAssets)
+}
+
+func (s *assetsSuite) TestUpdateObserverCanceledUnhappyCacheStillProceeds(c *C) {
+	// make sure that trying to remove the file from cache will no break the
```
Comment (anonymouse64) [inline boot/assets_test.go:1589]:
```suggestion
	// make sure that trying to remove the file from cache will not break the
```

**Pull request merged**